2.9BSD/usr/man/cat2/reboot.2

REBOOT(2)           UNIX Programmer's Manual            REBOOT(2)

NAME
     reboot - reboot system or halt processor

SYNOPSIS
     #include <sys/types.h>
     #include <sys/reboot.h>

     reboot(howto, dev)
     int howto;
     dev_t dev;

DESCRIPTION
     _R_e_b_o_o_t is used to cause a system reboot, and is invoked
     automatically in the event of unrecoverable system failures.
     _H_o_w_t_o is a mask of options passed to the bootstrap program.
     When none of these options is given, the system is rebooted
     from file unix in the file system specified by _d_e_v.
     Currently, this option is ignored, and the device used is
     the current root file system.  An automatic consistency
     check of the disks is then normally performed.

     The bits of _h_o_w_t_o are, in the order interpreted:

     RB_NOSYNC
          No sync (see _s_y_n_c(2)) is to be done.

     RB_DUMP
          A dump of memory is produced on the system's dump dev-
          ice, normally in swap space, as if the system had done
          a panic.

     RB_HALT
          The processor is simply halted; no reboot takes place.
          This should be used with caution.

     RB_ASKNAME
          Interpreted by the bootstrap program itself, causing it
          to inquire as to what file should be booted.  Normally,
          the system is booted from the file xx(0,0)unix without
          asking, where _x_x corresponds to the root file system
          device.

     RB_SINGLE
          Normally, the reboot procedure involves an automatic
          disk consistency check and then multi-user operations.
          This prevents the consistency check, rather simply
          booting the system with a single-user shell on the con-
          sole.  This switch is interpreted by the _i_n_i_t(8) pro-
          gram in the newly booted system.

     RB_NOFSCK
          The normal file system consistency check is omitted

Printed 5/22/83                                                 1

REBOOT(2)           UNIX Programmer's Manual            REBOOT(2)

          after the reboot.  This is used for a fast reboot on a
          quiet system with no possibility of file system prob-
          lems.

ERRORS
     _R_e_b_o_o_t will fail if:

     [EPERM]             The process's effective user ID is not
                         the super-user.

SEE ALSO
     crash(8), init(8), reboot(8)

BUGS
     The device specification is not implemented yet.

Printed 5/22/83                                                 2



2.9BSD/usr/src/cmd/reboot.c
static	char *sccsid = "@(#)reboot.c	4.5 (Berkeley) 4/21/81";
/*
 * Reboot
 *   Required signal handling:
 *	/etc/init must go to idle mode when sent a SIGQUIT.
 *	A subsequent SIGHUP sends init back to normal mode.
 *	The kernel must exempt the current process from kill(-1,SIGKILL).
 *   The reboot system call must be implemented!
 */
#include <whoami.h>
#include <sys/param.h>
#include <stdio.h>
#include <sys/reboot.h>
#include <errno.h>
#include <signal.h>
#include <time.h>

#define SHUTDOWNLOG "/usr/adm/shutdownlog"
#define DEVDIR "/dev/"
#define	KMEM	"/dev/kmem"

int halt;

main(argc, argv)
	int argc;
	char **argv;
{
	register howto;
	register char *argp;
	register i;
	int qflag = 0;
	dev_t	bootdev=NODEV, getdev(), rootdev();

	argc--, argv++;
	howto = 0;
	while (argc > 0) {
		if (!strcmp(*argv, "-q"))
			qflag++;
		else if (!strcmp(*argv, "-n"))
			howto |= RB_NOSYNC;
		else if (!strcmp(*argv, "-a"))
			howto |= (RB_ASKNAME | RB_SINGLE);
		else if (!strcmp(*argv,"-f"))
			howto |= RB_NOFSCK;
		else if (!strcmp(*argv,"-d"))
			howto |= RB_DUMP;
		else if (!strcmp(*argv,"-h")) {
			howto |= RB_HALT;
			halt++;
		} else if ((bootdev=getdev(*argv)) == NODEV) {
			fprintf(stderr,
	      "usage: reboot [ -n ][ -q ][ -f ][ -a ][ -d ][ -h ][ dev ]\n");
			exit(1);
		}
		argc--, argv++;
	}

	if (bootdev==NODEV)
		bootdev = rootdev();
	for (i = 1; i < NSIG; i++)
		signal(i, SIG_IGN);

	if (kill(1, SIGQUIT) == -1) {
		fprintf(stderr, "reboot: can't idle init\n");
		exit(1);
	}

	if (!qflag) for (i = 1; ; i++) {
		if (kill(-1, SIGKILL) == -1) {
			extern int errno;

			if (errno == ESRCH)
				break;

			perror("reboot: kill");
			kill(1, SIGHUP);
			exit(1);
		}
		if (i > 5) {
	fprintf(stderr, "CAUTION: some process(es) wouldn't die\n");
			break;
		}
		setalarm(2 * i);
		pause();
	}

	if ((howto & RB_NOSYNC) == 0)
		log_entry();
	if (!qflag) {
		if (!(howto & RB_NOSYNC)) {
			markdown();
			sync();
			sync();
		}
		setalarm(5);
		pause();
	}
	reboot(bootdev, howto);
	perror("reboot");
	/*
	 *  Reboot failed.  Tell init to go back to work.
	 */
	kill(1, SIGHUP);
	exit(1);
}

dingdong()
{
	/* RRRIIINNNGGG RRRIIINNNGGG */
}

setalarm(n)
{
	signal(SIGALRM, dingdong);
	alarm(n);
}

#include <utmp.h>
#define SCPYN(a, b)	strncpy(a, b, sizeof(a))
char	wtmpf[]	= "/usr/adm/wtmp";
struct utmp wtmp;

markdown()
{
	register f = open(wtmpf, 1);
	if (f >= 0) {
		lseek(f, 0L, 2);
		SCPYN(wtmp.ut_line, "~");
		SCPYN(wtmp.ut_name, "shutdown");
		time(&wtmp.ut_time);
		write(f, (char *)&wtmp, sizeof(wtmp));
		close(f);
	}
}

char *days[] = {
	"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"
};

char *months[] = {
	"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep",
	"Oct", "Nov", "Dec"
};

log_entry()
{
	FILE *fp;
	struct tm *tm, *localtime();
	time_t now;

	time(&now);
	tm = localtime(&now);
	fp = fopen(SHUTDOWNLOG, "a");
	if (fp == NULL)
		return;
	fseek(fp, 0L, 2);
	fprintf(fp, "%02d:%02d  %s %s %2d, %4d.  %s.\n", tm->tm_hour,
		tm->tm_min, days[tm->tm_wday], months[tm->tm_mon],
		tm->tm_mday, tm->tm_year + 1900,
		halt? "Halted": "Shutdown for reboot");
	fclose(fp);
}

#include <sys/stat.h>
char dev[2*DIRSIZ] = DEVDIR;
/*
 * Figure out the dev for a given string, e.g. "hp0a".
 * Returns NODEV if device doesn't exist, or isn't a block device.
 */
dev_t
getdev(s)
char *s;
{
	struct stat statbuf;
	char *index();

	if (index(s,'/')==NULL) {
		strcat(dev,s);
		s = dev;
	}
	if (stat(s,&statbuf) == -1)
		return(NODEV);
	if ((statbuf.st_mode&S_IFMT) != S_IFBLK)
		return(NODEV);
	return(statbuf.st_rdev);
}

/*
 * get the dev for the root filesystem.
 */
dev_t
rootdev() {
	struct stat statbuf;

	if (stat("/",&statbuf) == -1) {
		fprintf(stderr,"Can't stat root\n");
		exit(1);
	}
	return(statbuf.st_dev);
}
